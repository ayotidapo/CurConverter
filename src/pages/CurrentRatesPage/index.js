import { useEffect } from 'react';
import styled from 'styled-components';
import Input from 'components/Input';
import currencies from 'helpers/currencies';
import { useDispatch, useSelector } from 'react-redux';
import Loader from 'components/Loader';
import { useFormik } from 'formik';
import { getAllRates } from 'redux/actions/currentPageAction';
import { formatAmount } from 'helpers/functions';

const RatesWrapper = styled.div`
  width: 85%;

  margin: 0 auto;
  margin-bottom: 20px;

  .rates_dv {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(120px, 1fr));
    padding: 10px;
    overflow-y: scroll;
    text-align: center;
    cursor: pointer;
    border: 1px solid #e5e5e5;
    height: calc(100vh - 150px);
  }
  select {
    min-width: 200px;
    width: 200px;
    border: 1px solid var(--pri-color);
  }
  .goBck {
    cursor: pointer;
    color: var(--pri-color);
  }
`;

const CurrentRates = (props) => {
  const formik = useFormik({
    initialValues: {
      base: '',
    },
  });

  const { values, handleChange } = formik;

  const dispatch = useDispatch();
  const allRates = useSelector((state) => state.allRatesState);
  const { loading, data } = allRates;
  const historyState = props.history.location.state;
  const base = values?.base || historyState.base;

  useEffect(() => {
    dispatch(getAllRates(base));
  }, [base]);

  return (
    <RatesWrapper>
      <span className="goBck" onClick={() => props.history.goBack()}>
        Go Back
      </span>
      <Input
        data={currencies}
        defaultValue={base}
        id="base"
        name="base"
        label=""
        onChange={handleChange}
        type="select"
      />
      <div className="rates_dv">
        {loading ? (
          <Loader />
        ) : (
          data.map((obj) => (
            <p key={obj.currency}>
              <strong>1 {obj.currency} </strong> = {formatAmount(obj.rate, 3)}{' '}
              &nbsp;
            </p>
          ))
        )}
      </div>
    </RatesWrapper>
  );
};

export default CurrentRates;
