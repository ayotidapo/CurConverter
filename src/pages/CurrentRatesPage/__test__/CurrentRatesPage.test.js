/* eslint-disable no-undef */
/* eslint-disable import/prefer-default-export */
import React from 'react';
import axiosMock from 'axios';
import history from 'helpers/history';
import { renderWithRedux } from 'testHelper';
import { cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';

import CurrentRatesPage from '../index';

describe('how currentRates Page component works', () => {
  let cProps;
  afterEach(cleanup);
  beforeEach(() => {
    cProps = {
      history: {
        ...history,
        location: {
          ...history.location,
          state: {
            base: 'USD',
          },
        },
      },
    };
    axiosMock.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          base: 'USD',
          results: {
            AED: 3.67299,
            AFN: 89.85526,
            ALL: 104.6767,
          },
        },
      }),
    );
  });
  it('should renders the component without crash and takes snapshot', async () => {
    const { asFragment } = renderWithRedux(<CurrentRatesPage {...cProps} />);
    expect(asFragment).toMatchSnapshot();
    expect(axiosMock.get).toHaveBeenCalledTimes(1);

    expect(axiosMock.get).toHaveBeenCalledWith(
      `/fetch-all?from=USD&api_key=${process.env.REACT_APP_CUR_CONV_KEY}`,
    );
  });

  it('should renders the loader while data changes and show  new content when dome', async () => {
    const { getByTestId } = renderWithRedux(<CurrentRatesPage {...cProps} />);
    const baseSelectInput = getByTestId('rates-base');

    const loader = getByTestId('loader');

    userEvent.selectOptions(baseSelectInput, 'USD');

    expect(loader).toBeVisible();
    expect(axiosMock.get).toHaveBeenCalledTimes(1);

    expect(axiosMock.get).toHaveBeenCalledWith(
      `/fetch-all?from=USD&api_key=${process.env.REACT_APP_CUR_CONV_KEY}`,
    );

    const contentElement = await screen.findByTestId('content-div');
    expect(contentElement).toBeInTheDocument();
  });
});
