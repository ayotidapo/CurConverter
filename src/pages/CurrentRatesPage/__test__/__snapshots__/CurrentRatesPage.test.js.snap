// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`how currentRates Page component works renders the component without crash and takes snapshot 1`] = `[Function]`;
